
## 📁 Repository Structure

The repository is organized into different  folders, each corresponding to one or more figures in the manuscript:

| Folder | Description | Related Figure(s) |
|--------|-------------|-------------------|
| `bulkRNAseq/`        | Scripts and metrics related to bulk-RNA seq | Figure 5 |
| `cfdna-mutation-overlap/`     | For looking into mutation capture in cell-free DNA based on matched tumors | Figure 7 |
| `cfdna-nucleosome-profiling`     | Script to produce the heatmap for TFs from nucleosome profiling results| Figure 7 |
| `comut_mutations_cna_sv/`| Landscape plot for muttaions, CNAs and SVs. | Figure 3 |
| `copy-number-tumors/`    | Copy number alterations and founder, shared, and private clonality statys  | Figure 3 |
| `experimental_analysis/`      | Fanconi Anemia pathway analysis and experimental vadliation and assays | Figure 3 |
| `metastatic-seeding/`      | Metastatic seeding runs from MACHINA and downstream analysis | Figures 1 and 2 |
| `mutational-signatures/`      | Mutational signatures analysis and clonality | Figure 4 |
| `pathogenic_alteration_analysis/`      | Analysis for combining different alteration types such as mutations, CNAs and svs | Figure 3 |
| `pyclone-vi_lichee/`      | The basis for phylogenetic trees and clonality for mutations | Figures 1, 2 ,3 and 7 |
| `simple-SV/`      | Consensus and filtering for looking at simple structural variants such as insertions, deletions, duplications, inversions and translocations| Figure 3 |
| `snRNAseq/`      | Single-cell analysis including clustering and Nichenet | Figure 6 |
| `svclone/`      | Demo run and downstream analysis for svclone | Figure 3 |
v


> 📌 **Note**: Figures may be supported by multiple scripts and cross-reference data from other folders.
